Package cannot be imported for certain TypeScript configurations

**Describe the bug**

I apologize in advance for filing this; ESM/CJS/TypeScript compatibility is a horrific can of worms.

Importing this package in a TypeScript project that uses `"module": "Node16"` fails when the project itself is not configured to use ESM (that is, `package.json` does not contain `"type": "module"`).

I've created a reproduction here: https://github.com/nwalters512/sql-formatter-typescript-repro. Instructions are in the readme.

**Expected behavior**

Given that `sql-formatter` provides both ESM and CJS implementations, I would expect to be able to import the CJS version in this situation.

**Actual behavior**

TypeScript produces the following error:

```
src/index.ts:1:24 - error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("sql-formatter")' call instead.
  To convert this file to an ECMAScript module, change its file extension to '.mts', or add the field `"type": "module"` to '/Users/nathan/git/sql-formatter-typescript-repro/package.json'.

1 import { format } from 'sql-formatter';
                         ~~~~~~~~~~~~~~~


Found 1 error in src/index.ts:1
```

Note that things still work correctly at runtime; the presence of the `types` field combined with `"type": "module"` in this package confused TypeScript enough that it doesn't realize that there is a completely valid CJS implementation. I considered filing an issue upstream on TypeScript itself, but this was already discussed on https://github.com/microsoft/TypeScript/issues/54263, which was closed without any indication that TypeScript itself would fix this (even though I think they should).

I believe the way to fix this would be to drop the `"types"` field from `"exports"` and instead place `*.d.ts`/`*.d.cts` files adjacent to the files that are referenced by the `"import"` and `"require"` exports in `package.json`. This way, TypeScript will automatically discover the correct types based on the particular conditions that it's using to resolve the implementation files.

--- a/bin/sql-formatter-cli.cjs
+++ b/bin/sql-formatter-cli.cjs
@@ -2,7 +2,7 @@
 
 'use strict';
 
-const { format, supportedDialects } = require('../dist/sql-formatter.min.cjs');
+const { format, supportedDialects } = require('../dist/index.cjs');
 const fs = require('fs');
 const tty = require('tty');
 const { version } = require('../package.json');

--- a/package.json
+++ b/package.json
@@ -3,16 +3,14 @@
   "version": "14.0.0",
   "description": "Format whitespace in a SQL query to make it more readable",
   "license": "MIT",
-  "main": "dist/sql-formatter.min.cjs",
+  "main": "dist/index.cjs",
+  "module": "dist/index.js",
   "unpkg": "dist/sql-formatter.min.js",
-  "module": "lib/index.js",
-  "types": "lib/src/index.d.ts",
   "exports": {
     "./package.json": "./package.json",
     ".": {
-      "types": "./lib/src/index.d.ts",
-      "import": "./lib/index.js",
-      "require": "./dist/sql-formatter.min.cjs"
+      "import": "./dist/index.js",
+      "require": "./dist/index.cjs"
     }
   },
   "bin": {
@@ -64,10 +62,9 @@
     "prepare": "yarn clean && yarn grammar && yarn fix && yarn check && yarn build",
     "pre-commit": "npm-run-all --parallel ts:changes lint:changes",
     "grammar": "nearleyc src/parser/grammar.ne -o src/parser/grammar.ts",
-    "build:babel": "babel src --out-dir lib --extensions .ts --source-maps",
-    "build:types": "tsc --emitDeclarationOnly --isolatedModules",
-    "build:minified": "webpack --config webpack.prod.js && cp dist/sql-formatter.min.cjs dist/sql-formatter.min.js",
-    "build": "yarn grammar && npm-run-all --parallel build:babel build:types build:minified",
+    "build:tsup": "tsup src/index.ts --format cjs,esm --sourcemap --dts",
+    "build:webpack": "webpack --config webpack.prod.js && cp dist/sql-formatter.min.cjs dist/sql-formatter.min.js",
+    "build": "yarn grammar && npm-run-all --parallel build:tsup build:webpack",
     "release": "release-it"
   },
   "repository": {
@@ -113,6 +110,7 @@
     "rimraf": "^3.0.2",
     "ts-jest": "^28.0.5",
     "ts-loader": "^9.3.1",
+    "tsup": "^8.0.1",
     "typescript": "^4.7.4",
     "webpack": "^5.74.0",
     "webpack-cli": "^4.9.1",

--- a/src/index.ts
+++ b/src/index.ts
@@ -2,9 +2,24 @@ export { supportedDialects, format, formatDialect } from './sqlFormatter.js';
 export { expandPhrases } from './expandPhrases.js';
 export { ConfigError } from './validateConfig.js';
 
-// Intentionally use "export *" syntax here to make sure when adding a new SQL dialect
-// we wouldn't forget to expose it in our public API.
-export * from './allDialects.js';
+// When adding a new dialect, be sure to add it to the list of exports below.
+export { bigquery } from './languages/bigquery/bigquery.formatter.js';
+export { db2 } from './languages/db2/db2.formatter.js';
+export { db2i } from './languages/db2i/db2i.formatter.js';
+export { hive } from './languages/hive/hive.formatter.js';
+export { mariadb } from './languages/mariadb/mariadb.formatter.js';
+export { mysql } from './languages/mysql/mysql.formatter.js';
+export { n1ql } from './languages/n1ql/n1ql.formatter.js';
+export { plsql } from './languages/plsql/plsql.formatter.js';
+export { postgresql } from './languages/postgresql/postgresql.formatter.js';
+export { redshift } from './languages/redshift/redshift.formatter.js';
+export { spark } from './languages/spark/spark.formatter.js';
+export { sqlite } from './languages/sqlite/sqlite.formatter.js';
+export { sql } from './languages/sql/sql.formatter.js';
+export { trino } from './languages/trino/trino.formatter.js';
+export { transactsql } from './languages/transactsql/transactsql.formatter.js';
+export { singlestoredb } from './languages/singlestoredb/singlestoredb.formatter.js';
+export { snowflake } from './languages/snowflake/snowflake.formatter.js';
 
 // NB! To re-export types the "export type" syntax is required by webpack.
 // Otherwise webpack build will fail.

